Application: Shopbring

zhiquan911 opened the pull request:
# Grant Application Checklist

- [x] The [application-template.md](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( "project_name.md") and updated.
- [x] A BTC address for the payment of the milestones is provided inside the application.  
- [x] The software of the project will be released under the Apache license version 2.0 as specified in the [terms and conditions](https://gist.github.com/Noc2/75bc58e8ce9b5d419ff883b0cf2b8c19).
- [x] The total funding amount of the project is below CHF 30k at the time of submission.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/w3f/Open-Grants-Program?pullRequest=158) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/w3f/Open-Grants-Program?pullRequest=158) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/w3f/Open-Grants-Program?pullRequest=158) it.</sub>

Noc2 reviewed (changes requested):
Thanks for the application. Sounds like an interesting project. I have a few questions/comments:

- Could clarify a little bit better how you deal with private data? So you don’t put any private data on-chain encrypted or unencrypted, correct?   
- Do you already have some UI mock-ups/designs which you can share? 
- It might make sense to initially only apply for funding for milestone 1 and 2. On the one hand parachains are not live yet, and on the other hand you can potentially reuse some of the other oracle modules that are currently under development. So could you for now only focus on milestone 1 and 2 with your application?

zhiquan911 commented:
> * Could clarify a little bit better how you deal with private data? So you don’t put any private data on-chain encrypted or unencrypted, correct?
> * Do you already have some UI mock-ups/designs which you can share?
> * It might make sense to initially only apply for funding for milestone 1 and 2. On the one hand parachains are not live yet, and on the other hand you can potentially reuse some of the other oracle modules that are currently under development. So could you for now only focus on milestone 1 and 2 with your application?

**Reply to Question 1**

We will use [ECDH](https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman) to establish a shared secret for both parties in the commissioned shopping process. We will use [SM2(GM/T 0003-2012)](https://zh.wikipedia.org/wiki/SM2) for the program implementation.

The preliminary design is:

1. The comissioned shopping initiator uses the current nonce to derive a 256-bit key and its SM2-publicKey, and record SM2-publicKey on the commissioned shopping order.
1. The shopping agent also uses the current nonce to derive a 256bit key and its SM2-publicKey when accepting the commissioned shopping order, and record SM2-publicKey on this order.
1. In this way, both parties build a shared secret through the SM2 key agreement algorithm, which is used to encrypt the privacy information in the order.

**Reply to Question 2**

Sorry, UI mock-ups/designs is not yet available. Our project's UI design is still in draft.

**Reply to Question 3**

Yes, we plan to complete the development of milestone 1 and 2 between January and March 2021. Polkadot decoded plan is to release Rococo v1 in December 2020, and parachains may be launched on Kusama in March 2021.
Around that time, we were able to carry out milestone 3.

Noc2 commented:
Thanks for the quick reply. Could you integrate this information into the application? So are you storing the encrypted data on-chain or are you also using an off-chain storage solution? Also could you still remove milestone 3, at least the “Parachain Testnet Debugging” part.

zhiquan911 reviewed (commented):
> Thanks for the quick reply. Could you integrate this information into the application? So are you storing the encrypted data on-chain or are you also using an off-chain storage solution? Also could you still remove milestone 3, at least the “Parachain Testnet Debugging” part.

I have added the data encryption idea to the application.
The order's privacy data will be stored on-chain.
I have changed the content of milestone 3. We will refer to Cumulus Workshop for the parachain interoperability test in the local environment.

Noc2 reviewed (changes requested):
Sorry for the late reply. It seems I missed your update. Could you update the BTC price or choose DAI as the payment option? Storing the private data on-chain (even encrypted) is from a security perspective a big issue. Keep in mind this data is stored there indefinitely (GDPR) and generally you shouldn’t store a lot of data on-chain (since it’s quite expensive). Also I still believe we wouldn’t support “Parachain Local Testing” or milestone 3 in general. The local testing is quite specific to your own solution. And for the off-chain worker oracle solution, we would need to have a lot more details (there are quite a few teams currently working on solutions). Furthermore could you add the programming language to the deliverables as well as more information about the functionality of the modules? For example, why do you need a “Payment Currency” and what functions would it have?

zhiquan911 commented:
> Sorry for the late reply. It seems I missed your update. Could you update the BTC price or choose DAI as the payment option? Storing the private data on-chain (even encrypted) is from a security perspective a big issue. Keep in mind this data is stored there indefinitely (GDPR) and generally you shouldn’t store a lot of data on-chain (since it’s quite expensive). Also I still believe we wouldn’t support “Parachain Local Testing” or milestone 3 in general. The local testing is quite specific to your own solution. And for the off-chain worker oracle solution, we would need to have a lot more details (there are quite a few teams currently working on solutions). Furthermore could you add the programming language to the deliverables as well as more information about the functionality of the modules? For example, why do you need a “Payment Currency” and what functions would it have?

Thank you for your reply. I have updated the document and the payment option has been changed to DAI.

I accept your opinion. I now save the privacy data of order details in the off-chain order system, and only the order hash is recorded on chain.

`About off-chain worker oracle solution`. We now only have invoices that need to be verified off chain, so the system introduces the role of inspector, which provides API server for verifying invoices. The inspector needs to run the Shopbring node. And the off-chain worker will read the invoice hash that has not yet been verified, request the inspector's API server. This server reads the invoice information of the off-chain order system for verification, and the inspector will submit the verification result to the chain.

I deleted milestone 3 and added the functions to be developed by each module.

Noc2 commented:
Thanks for the update. The application looks much better. But we wouldn’t support the “UI Design”, only the actual development of the UI in case it’s open source. Also there is already a [generic asset module](https://github.com/paritytech/substrate/tree/v2.0.0-rc6/frame/generic-asset). How is your solution different from the above. Could you update this and potentially reduce the price accordingly.

zhiquan911 commented:
> Thanks for the update. The application looks much better. But we wouldn’t support the “UI Design”, only the actual development of the UI in case it’s open source. Also there is already a [generic asset module](https://github.com/paritytech/substrate/tree/v2.0.0-rc6/frame/generic-asset). How is your solution different from the above. Could you update this and potentially reduce the price accordingly.

I deleted "UI Design". And reduced development costs.

Because our multiple assets come from other parachain cross-chain transfers, we need to make the following modifications to the `generic asset module`:

1. Asset registration shall be called by the proposal of the council.
2. The issuance and redemption of assets must be triggered by cross-chain transfers.

Noc2 commented:
> Because our multiple assets come from other parachain cross-chain transfers, we need to make the following modifications to the `generic asset module`:
> 
> 1. Asset registration shall be called by the proposal of the council.
> 2. The issuance and redemption of assets must be triggered by cross-chain transfers.

Could you integrate this into the application and also integrate unit tests into milestone1? Also could you provide more information regarding “Implement Shopbring App”, e.g. programming language. Do you already have some mockups, which you can share? Furthermore the numbers of milestone 2 seem to be off.

zhiquan911 commented:
> Could you integrate this into the application and also integrate unit tests into milestone1? Also could you provide more information regarding “Implement Shopbring App”, e.g. programming language. Do you already have some mockups, which you can share? Furthermore the numbers of milestone 2 seem to be off.

Thank you for your reply. I have updated the application according to your requirements.

- The content of **milestone** has been updated.
- **Additional Information** also added the latest UI mock-ups demo link. You can also click [here](https://modao.cc/app/d7520aeb5b9a99c55e3e49e39e19b5cc736e3d3d?simulator_type=device&sticky) to view.

zhiquan911 commented:
@Noc2 
Hi, what else do I need to change, please point out?

Noc2 reviewed (approved):
Sorry for the late reply and thanks for the update. Looks good to me. Maybe you can also leverage IPFS at some stage for the off-chain order system (but it would also raise additional privacy concerns). I will share the application with the rest of the team.

mmagician commented on applications/Shopbring.md:
```diff
@@ -0,0 +1,320 @@
+| 0b. | Documentation | We write complete program flow description and system module interface definition. |
+| 0c. | Testing Guide | We will deploy a POA consensus Shopbring network, which can call system modules through polkadot.js. |
+| 0d. | Article/Tutorial | We will write an article or tutorial that explains the work done as part of the grant. |
+| 1. | Substrate module: Generic Asset | Programming language: `Rust`. This module aims to manage multiple cross-chain assets. It has the following functions: `register_generic_assets`, `deregister_generic_assets`, `mint_generic_assets`, `burn_generic_assets`, `transfer_assets`. |
```
Have a look at the [Generic Token Dealer Pallet](https://github.com/subdarkdex/pallet-generic-token-dealer), perhaps you could re-use their work and make your lives easier :)

zhiquan911 replied:
Thank you, let me compare which is more suitable.

mmagician commented on applications/Shopbring.md:
```diff
@@ -0,0 +1,320 @@
+| 1. | Substrate module: Generic Asset | Programming language: `Rust`. This module aims to manage multiple cross-chain assets. It has the following functions: `register_generic_assets`, `deregister_generic_assets`, `mint_generic_assets`, `burn_generic_assets`, `transfer_assets`. |
+| 2. | Substrate module: Invitation | Programming language: `Rust`. This module aims to establish the invitation relationship between accounts. It has the following functions: `register_inviter`, `accept_invitation`, `end_invitation_period`. |
+| 3. | Substrate module: Commissioned Shopping | Programming language: `Rust`. This module aims to implement the commissioned shopping process. It has the following functions: `apply_shopping_order`, `accept_shopping_order`, `do_commodity_shipping`, `confirm_commodity_received`, `apply_commodity_return`, `accept_commodity_return`, `do_commodity_returning`, `confirm_commodity_returned`. |
+| 4. | Substrate module: Transaction Payment | Programming language: `Rust`. This module aims to rewrite the `pallet_transaction_payment` so that it can change the payment `origin` of the transaction fee according to the specific module's functions.|
```
Does it mean you're planning to fork and maintain your own pallet, or rather do you aim to submit a PR to merge these changes upstream? With the details provided, it is difficult to judge whether one or the other option is preferable - though generally maintaining your own fork is discouraged.

zhiquan911 replied:
In the early stage, we intend to fork and maintain our own pallet, which can avoid some version conflicts. When we are free in later work, we will submit a PR to merge these changes upstream.

zhiquan911 commented on applications/Shopbring.md:
```diff
@@ -0,0 +1,320 @@
+
+Normal user registration arbitrators need to freeze a certain native token as an application condition. The `council` will review the applicant's information, and more than half of the members will vote to become an arbitrator.
+
+The arbitrator’s obligation is that when a common user has a dispute during the commissioned shopping, either party will appeal to the system. The arbitrator can choose the case to intervene in the dispute. During the waiting period, up to `M` arbitrators can sign up to participate in the dispute. After the waiting period ends, the system will randomly select `N`($N < M$, and `N` is an odd number ...
```
In order to establish a good arbitration mechanism in the early stage of the network, the council will consider people with customer service experience to be arbitrators when reviewing applications. It is also considered that there are not many commissioned shopping disputes in the early stage of the network, so the wages of the arbitrator will be fixed and paid by the treasury. This can reduce the cost of both parties to the transaction and encourage more arbitrators to actively participate in dispute resolution.

When the network's credit incentive mechanism is perfected, all accounts with credit value have been rewarded with native tokens. At this time, both parties need to deposit a certain native tokens (also can be deposited by other account) for commissioned shopping. Once a dispute occurs during the process, the party who loses the arbitration will pay the arbitrators.

zhiquan911 commented on applications/Shopbring.md:
```diff
@@ -0,0 +1,320 @@
+
+Normal user registration arbitrators need to freeze a certain native token as an application condition. The `council` will review the applicant's information, and more than half of the members will vote to become an arbitrator.
+
+The arbitrator’s obligation is that when a common user has a dispute during the commissioned shopping, either party will appeal to the system. The arbitrator can choose the case to intervene in the dispute. During the waiting period, up to `M` arbitrators can sign up to participate in the dispute. After the waiting period ends, the system will randomly select `N`($N < M$, and `N` is an odd number ...
```
In the early stage of the network, there may be fewer arbitrators. When the dispute does not have enough arbitrators to intervene, the commissioned shopping cannot continue. At this time, the council has the responsibility to access.

Arbitrators can get wages paid by the treasury when they intervene in disputes. When there are more disputes, they will naturally attract more arbitrators to join the network to actively intervene in disputes, so that the council has very little opportunity to intervene in disputes.

zhiquan911 commented on applications/Shopbring.md:
```diff
@@ -0,0 +1,320 @@
+
+#### The Commissioned Shopping Process
+
+> Users continue to make the commissioned shopping on the platform, and the system will record considerable order data. If the order data is stored on chain, it will incur very large storage costs. Therefore, we will develop an off-chain order management system to store order detail, like shopping order information, shipping information, invoice information, and return information. These informat ...
```
The current development milestones have not considered decentralized storage solutions. Our off-chain order system is mainly used to store detailed order data. The system only has the logic for account authorization and the logic for querying account-related order data.

We know that polkadot has some teams working on decentralized storage solutions, so we don't need to spend a lot of time studying this aspect. When the decentralized storage solution on polkadot is available for production, we will consider data migration.

zhiquan911 commented on applications/Shopbring.md:
```diff
@@ -0,0 +1,320 @@
+
+* Transaction fees: A portion of each block's transaction fees goes to the Treasury, with the remainder going to the block author.
+* The commissioned shopping tip: A portion of each commissioned shopping transaction tip goes to the Treasury.
+* Native token inflation: As the credit value of the entire network increases, native tokens are also issued.
```
The issuance mechanism of native currency is still being conceived. The preliminary idea is:

We will issue 1 million native tokens in the genesis block, part of which will be allocated to seed investors and venture capitalists, and part of them will be used for community promotion and establish an inviter relationship.

The system initially sets an equation for credit value and native token value. Assumption: 1 credit value = 0.001 native token.

After the user completes the verification of the shopping receipts, the account credit value can be increased. The program will calculate the number of native tokens obtained by the account based on the newly added credit value and record it in the reward mapping table. When users receive rewards, a portion of the rewards are donated to the treasury.

zhiquan911 commented on applications/Shopbring.md:
```diff
@@ -0,0 +1,320 @@
+Shopbring introduces the role of inspector to verify whether the shopping process actually happened. This role provides trusted services to verify shopping receipts(like invoices). For example, in China, most e-commerce platforms provide the right of applying for electronic invoices issued by the State Administration of Taxation. By verifying the authenticity of the electronic invoice, it can be  ...
+
+* The user submits the application for registering inspector, then waits for the council to review. When more than half of the members of the council vote for, the review is passed.
+* The inspector needs to provide an API service to verify the shopping receipt. Shopbring's off-chain worker will submit the unverified shopping receipt to the inspector's API service, and finally the inspector will submit the verification result to the chain.
```
**We define the structure of shopping receipt like this：**

| Parameter      | Type     | Description                                                                                   |
|----------------|----------|-----------------------------------------------------------------------------------------------|
| area           | u32      | Country or area code.                                                                         |
| receipt_type   | u8       | Receipt type: 0: e-invoice, 1: physical bill, need to be photographed and uploaded.           |
| payload        | Vec\<u8> | E-invoice information is saved in json format according to different national standards.      |
| receipt_attach | Vec\<u8> | The link to the physical bill attachment requires manual verification.                        |
| receipt_hash   | Vec\<u8> | The receipt's data will be encoded using SCALE-codec and finally be hashed using blake2b-256. |
| order_hash     | Vec\<u8> | Order hash, used to associate original order detail.                                         |

1. The registered inspector can bind his verification API on-chain.
2. Because there may be many types of receipt formats, we store the detail of the receipt off-chain, and after encoding the data, submit the `receipt hash` to the chain for storage.
3. The verification system of the inspector needs to connect to the off-chain order system, and send the `receipt hash` to it to fetch the detail of the receipt. The verification system completes verification of the receipt data and records the results.
4. The off-chain worker sends the `receipt hash` to the verification API, fetchs the verification result, and submits the transaction signed by the inspector account to the chain.

zhiquan911 commented on applications/Shopbring.md:
```diff
@@ -0,0 +1,320 @@
+Each stage has an operation time limit. If the time-out occurs, there are the following processing conditions:
+
+* No shopping agents timed out in response to the commissioned shopping order, the order will be closed automatically, and the locked payment amount and tip will be returned to the consumer.
+* When the consumer exceeds the time limit for submitting the delivery address, the order is automatically closed, and the order tip is paid to the shopping agent.
```
Yes, I will simplify this process.

zhiquan911 commented:
@mmagician @Noc2
Thank you for your questions, I have replied to the above questions. 
At the same time, I also made relevant amendments to the application.

zhiquan911 commented:
@Noc2 @mmagician 
Hi, what else do I need to change, please point out?

mmagician commented:
@zhiquan911 Unfortunately, your application hasn't convinced enough members of the Grants Committee and hasn't received the necessary approvals.
Thank you again for all the effort you put into this & we encourage you to apply with another idea in the future!

zhiquan911 commented:
@mmagician 
Thank you for your careful review. Although I didn't get grants, it inspired me more ideas in this process.
